### ISSUE TITLE
mozdownload sometimes fails on Taskcluster and Travis

### ISSUE THREAD

**Issue by foolip** [web-platform-tests/wpt#13274]
At least twice some of the Firefox tasks for pushes to master have failed in mozdownload:

https://tools.taskcluster.net/groups/AGupAeh6TrSdyNlDxcqZXw (for commit 2df7f9ff620cbdaa2928464892fb1dfb880fd6c6, now only discoverable via [API](https://api.github.com/repos/web-platform-tests/wpt/commits/2df7f9ff620cbdaa2928464892fb1dfb880fd6c6/statuses))
https://tools.taskcluster.net/groups/eRtqwYxjTfeob0g4hHcOlw (for commit 91491deb7dcfe4a0d0ece61f086345abad4d46a4)

The most recent failure was:
```
Traceback (most recent call last):
  File "./wpt", line 5, in <module>
    wpt.main()
  File "/home/test/web-platform-tests/tools/wpt/wpt.py", line 129, in main
    rv = script(*args, **kwargs)
  File "/home/test/web-platform-tests/tools/wpt/run.py", line 510, in run
    **kwargs)
  File "/home/test/web-platform-tests/tools/wpt/run.py", line 488, in setup_wptrunner
    kwargs["binary"] = setup_cls.install(venv, channel=channel)
  File "/home/test/web-platform-tests/tools/wpt/run.py", line 165, in install
    return self.browser.install(venv.path, channel)
  File "/home/test/web-platform-tests/tools/wpt/browser.py", line 134, in install
    destination=dest).download()
  File "/home/test/web-platform-tests/_venv/lib/python2.7/site-packages/mozdownload/factory.py", line 121, in __init__
    scraper_types[scraper_type].__init__(self, **kwargs)
  File "/home/test/web-platform-tests/_venv/lib/python2.7/site-packages/mozdownload/scraper.py", line 346, in __init__
    Scraper.__init__(self, *args, **kwargs)
  File "/home/test/web-platform-tests/_venv/lib/python2.7/site-packages/mozdownload/scraper.py", line 135, in __init__
    self._retry_check_404(self.get_build_info)
  File "/home/test/web-platform-tests/_venv/lib/python2.7/site-packages/mozdownload/scraper.py", line 150, in _retry_check_404
    self._retry(func, **retry_kwargs)
  File "/home/test/web-platform-tests/_venv/lib/python2.7/site-packages/mozdownload/scraper.py", line 141, in _retry
    return redo.retry(func, **retry_kwargs)
  File "/home/test/web-platform-tests/_venv/lib/python2.7/site-packages/redo/__init__.py", line 162, in retry
    return action(*args, **kwargs)
  File "/home/test/web-platform-tests/_venv/lib/python2.7/site-packages/mozdownload/scraper.py", line 403, in get_build_info
    self.date, self.build_index)
  File "/home/test/web-platform-tests/_venv/lib/python2.7/site-packages/mozdownload/scraper.py", line 484, in get_build_info_for_date
    raise errors.NotFoundError(message, url)
mozdownload.errors.NotFoundError: Folder for builds on 2018-09-28-22-04-33 has not been found: https://archive.mozilla.org/pub/firefox/nightly/2018/09/
```

The previous was very similar except for the date: "mozdownload.errors.NotFoundError: Folder for builds on 2018-09-24-10-03-54 has not been found: https://archive.mozilla.org/pub/firefox/nightly/2018/09/"

That it's happened 4 days apart suggests that it wasn't just a transient problem with archive.mozilla.org.

Here's where the error is thrown:
https://github.com/mozilla/mozdownload/blob/866cfebe9b8137bfe7ba8411efbe9d0e9d24093a/mozdownload/scraper.py

@jgraham, can you take a look?

**Comment by foolip** [comment]
The reason I noticed is that in https://wpt.fyi/test-runs?label=taskcluster, the row for 2df7f9ff62 has *only* stable results, but I expected stable + experimental. @Hexcles FYI.

**Comment by mdittmer** [comment]
@Hexcles could you please add a priority label to this? I think it's your call because you are managing priority/urgency of shipping TaskCluster runs on wpt.fyi.

**Comment by gsnedders** [comment]
https://github.com/web-platform-tests/wpt/commit/300f8e2c89abc98917b40e95c14f2edf1d354f99 failed with this on https://tools.taskcluster.net/groups/TKPIsqzARXyy88H0_n8bLw

**Comment by foolip** [comment]
@jgraham, can you take a look? Needs more retry?

**Comment by jgraham** [comment]
I fairly strongly suspect that this is happening when a new nightly is being released (maybe some platforms are available and some are not?). But we are already handling that badly; it's possible to end up with some tests run in the previous nightly and some in the new one. Really we need a single decsion task that picks a binary URL and makes it available to the subsequent tasks to ensure that they all run against the exact same version. Note that Chrome could have the same issue, but it's less likely since the releases are less often. But it's harder to solve in that case; we probably actually need to download the .deb and make it available as an artifact since there isn't a longlived URL AFAIK.

**Comment by foolip** [comment]
> we probably actually need to download the .deb and make it available as an artifact since there isn't a longlived URL AFAIK

Enter, stage left, @jugglinmike to say something about how this is done in https://github.com/web-platform-tests/results-collection

**Comment by foolip** [comment]
I happened to look at recent commits, and https://github.com/web-platform-tests/wpt/commits/75b92bf3d1791dc0e47cd8a716a135e98d2d2937 has a similar failure (https://tools.taskcluster.net/groups/ZklAzb_fTueVrBAWR0kGBA): "requests.exceptions.ConnectionError: HTTPSConnectionPool(host='hg.mozilla.org', port=443): Max retries exceeded with url: /mozilla-central/archive/tip.zip/testing/profiles/ (Caused by NewConnectionError('<urllib3.connection.VerifiedHTTPSConnection object at 0x7fc6f85775d0>: Failed to establish a new connection: [Errno 110] Connection timed out',))"

@jgraham is there anything we can do to mitigate this?

**Comment by foolip** [comment]
https://github.com/web-platform-tests/wpt/pull/14450#issuecomment-446135917 shows this happening when running unit tests in Travis too.

@jgraham, is there nothing we can do to make this more reliable?

**Comment by foolip** [comment]
This happened in https://github.com/web-platform-tests/wpt/pull/15012 now:
https://dev.azure.com/web-platform-tests/wpt/_build/results?buildId=4488

The failure was in test_install_firefox because of a network error in mozdownload:
https://gist.github.com/foolip/62ec6e42108ff93f22408340a84ac083#file-74-txt-L106

@jgraham should we mark the test as xfail?

**Comment by gsnedders** [comment]
@foolip that test shouldn't be any more flaky than any of the test run jobs where we start off by downloading the nightly browser

**Comment by foolip** [comment]
Perhaps it is one of few pytest tests that do this, because I've seen it fail a few times and I don't think I've seen it for other tests. But it happens for `./wpt run` of course, that was the beginning of this issue.

@jgraham is there an infra bug filed in some Mozilla repo for improving the reliability of this setup? Or could it be a mozdownload bug?

**Comment by foolip** [comment]
This happened for @annevk on https://github.com/web-platform-tests/wpt/pull/15122.

**Comment by jgraham** [comment]
It's a mozdownload bug. The way it works is that it looks for a directory containing builds, stores that, and then later uses the stored directory to look for the actual build. But the operation of creating a directory full of builds isn't (even nearly) atomic; the directory is created when the first artifacts are available not when the last build is complete. So the "solution" here is either a) stop using mozdownload and roll our own thing, b) rearchitect mozdownload to do the build and directory lookup in a single operation or c) catch the failure and try again with the previous build. c) is probably the most practical option but the tool really doesn't seem to be designed in a way that makes a fix here easy.

**Comment by gsnedders** [comment]
https://github.com/mozilla/mozdownload/issues/524 is the mozdownload bug, as appears above.

**Comment by foolip** [comment]
This happened on https://github.com/web-platform-tests/wpt/pull/15280.

@jgraham have you seen many Gecko exports blocked because of this? It seems from the stack that there's already retry involved, so I guess a fix for https://github.com/mozilla/mozdownload/issues/524 is the only hope?

**Comment by whimboo** [comment]
The latest stable nightly build could be found like this for linux64:
https://download.mozilla.org/?product=firefox-nightly-latest-ssl&os=linux64&lang=en-US

Which other kind of builds are required?

**Comment by foolip** [comment]
Those are the ones that have been failing, but we also download stable and beta for other runs. Not sure if those use mozdownload, @jgraham would know though.

**Comment by whimboo** [comment]
Well, beta will be https://download.mozilla.org/?product=firefox-beta-latest-ssl&os=linux64&lang=en-US.

**Comment by jgraham** [comment]
We can definitely experiment with not using mozdownload.

**Comment by jgraham** [comment]
Fixed via #15329

**Comment by foolip** [comment]
Thanks James!
